packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,442 @@
+  const plugins = getActivatedResourcePlugins(solutionSettings);
+  const envDefaultJson: { solution: { provisionSucceeded: boolean } } = await readJson(
+    path.join(fx, "env.default.json")
+  );
+  const hasFrontend = plugins?.some((plugin) => plugin.name === PluginNames.FE);
+  const hasBackend = plugins?.some((plugin) => plugin.name === PluginNames.FUNC);
+  const hasBot = plugins?.some((plugin) => plugin.name === PluginNames.BOT);
+  const hasProvision = envDefaultJson.solution?.provisionSucceeded as boolean;
+  return { hasFrontend, hasBackend, hasBot, hasProvision };
+}
+
+async function getMultiEnvFolders(projectPath: string): Promise<any> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  const fxConfig = path.join(fx, InputConfigsFolderName);
+  const templateAppPackage = path.join(projectPath, "templates", AppPackageFolderName);
+  const fxPublishProfile = path.join(fx, PublishProfilesFolderName);
+  await fs.ensureDir(fxConfig);
+  await fs.ensureDir(templateAppPackage);
+  return { fx, fxConfig, templateAppPackage, fxPublishProfile };
+}
+
+async function removeOldProjectFiles(projectPath: string): Promise<void> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  await fs.remove(path.join(fx, "env.default.json"));
+  await fs.remove(path.join(fx, "default.userdata"));
+  await fs.remove(path.join(fx, "settings.json"));
+  await fs.remove(path.join(fx, "local.env"));
+  await fs.remove(path.join(projectPath, AppPackageFolderName));
+  await fs.remove(path.join(fx, "new.env.default.json"));
+  // version <= 2.4.1, rmove .fx/appPackage.
+  await fs.remove(path.join(fx, AppPackageFolderName));
+}
+
+async function ensureProgrammingLanguage(
+  projectSettingPath: string,
+  envDefaultPath: string
+): Promise<void> {
+  const settings: ProjectSettings = await readJson(projectSettingPath);
+  if (!settings.programmingLanguage || !settings.defaultFunctionName) {
+    const envDefault = await readJson(envDefaultPath);
+    settings.programmingLanguage = envDefault[PluginNames.SOLUTION][programmingLanguage];
+    settings.defaultFunctionName = envDefault[PluginNames.FUNC][defaultFunctionName];
+    await fs.writeFile(projectSettingPath, JSON.stringify(settings, null, 4), {
+      encoding: "UTF-8",
+    });
+  }
+}
+
+async function getAppName(projectSettingPath: string): Promise<string> {
+  const settings: ProjectSettings = await readJson(projectSettingPath);
+  return settings.appName;
+}
+
+async function ensureActiveEnv(projectSettingPath: string): Promise<void> {
+  const settings: ProjectSettings = await readJson(projectSettingPath);
+  if (!settings.activeEnvironment) {
+    settings.activeEnvironment = "dev";
+    await fs.writeFile(projectSettingPath, JSON.stringify(settings, null, 4), {
+      encoding: "UTF-8",
+    });
+  }
+}
+
+async function cleanup(projectPath: string): Promise<void> {
+  const { _, fxConfig, templateAppPackage, fxPublishProfile } = await getMultiEnvFolders(
+    projectPath
```

Comment:
Also need to clean up the `local.env` file since it will no longer be used for local-debug if we enable multi-env. All local-debug related config will be stored in `localSettings.json`

Reply:
has remove it in `removeOldProjectFiles`